sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala

```diff
@@ -124,14 +124,9 @@ object ExpressionEncoder {
         s"`GetColumnByOrdinal`, but there are ${getColExprs.size}")
 
       val input = GetStructField(GetColumnByOrdinal(0, schema), index)
-      val newDeserializer = enc.objDeserializer.transformUp {
+      enc.objDeserializer.transformUp {
         case GetColumnByOrdinal(0, _) => input
       }
-      if (schema(index).nullable) {
-        If(IsNull(input), Literal.create(null, newDeserializer.dataType), newDeserializer)
```

Comment:
oh, I see. If the child deserializer is a tuple deserializer, it is just

```scala
val deserializer =
      NewInstance(cls, childrenDeserializers, ObjectType(cls), propagateNull = false)
```

So it misses the `If(IsNull(..), null, ...)` pattern. We should wrap the `NewInstance` with `If(IsNull(..), null)` at L139.

Reply:
good catch!